server/web.js

```diff
@@ -314,3 +381,427 @@ function getPluginVersion(callback) {
+    if (!threshold) {
+      threshold = 10000;
+    }
+    // TODO: Very large threshold => no results
+
+    // If there have been no changes to the document,
+    // ops = {v:0}
+    if (!Array.isArray(ops)) {
+      return [];
+    }
+
+    var chunkedDiffs = [];
+
+    /* Setup the baseline of the document */ 
+    var firstOp = ops[0];
+
+    // The baseline for the next diff
+    var currentBaseline = {v:0};
+    sharedb.ot.apply(currentBaseline, firstOp);
+    // The doc to apply ops to
+    var currentDoc = {v:0};
+    sharedb.ot.apply(currentDoc, firstOp);
+
+    var lastTs = firstOp.m.ts;
+
+    // Create a diff for the first part, so that
+    // we can track original code
+    var baseDiff = diff.diffLines(baseline.trim(), currentBaseline.data.text.trim());
+    baseDiff.forEach(function(part) {
+      // Note: should only be one part
+      part.original = true;
+    });
+
+    chunkedDiffs.push(baseDiff);
+
+    /* Apply each op, and calculate a diff if two 
+       consecutive ops are far enough apart */
+    for (var i = 1; i < ops.length; i++) {
+      var op = ops[i];
+
+      // Start a new chunk if necessary
+      if (op.m.ts - lastTs > threshold) {
+        var chunkedDiff = diff.diffLines(
+          currentBaseline.data.text.trim(), currentDoc.data.text.trim());
+        
+        // Only push diffs with changes
+        if (!(chunkedDiff.length == 1 && 
+            !chunkedDiff[0].added &&
+            !chunkedDiff[0].removed)) {
+          chunkedDiffs.push(chunkedDiff);
+        }
+
+        // Make a deep copy
+        currentBaseline = JSON.parse(JSON.stringify(currentDoc));
+        
+      }
+
+      // Apply the op
+      let err = sharedb.ot.apply(currentDoc, op);
+      if (err) {
+        // TODO: Better error handling
+        console.log("err when applying op:" + JSON.stringify(err));
+        return;
+      }
+         
+      lastTs = op.m.ts;
+    }
+
+    // Add the last diff
+    var chunkedDiff = diff.diffLines(
+      currentBaseline.data.text.trim(), currentDoc.data.text.trim());
+
+    // Only push diffs with changes
+    if (!(chunkedDiff.length == 1 &&
+        !chunkedDiff[0].added &&
+        !chunkedDiff[0].removed)) {
+      chunkedDiffs.push(chunkedDiff);
+    }
+
+    return chunkedDiffs; 
+}
+
+// TODO: Simplify
+// TODO: Remove parts with '' at the end
+
+/**
+ * Merges the given list of diffs into a total diff,
+ * maintaining the inserts and deletes that happened
+ * in each diff.
+ */
+function mergeDiffs(diffs) {
+  if (diffs.length == 0) {
+    return diffs;
+  }
+
+  mergedDiff = JSON.parse(JSON.stringify(diffs[0]));
+  for (var i = 1; i < diffs.length; i++) {
+    var diff = JSON.parse(JSON.stringify(diffs[i]));
+
+    // Index into mergedDiff for what chunk we're currently on
+    var currentChunkInMerged = 0;
+
+    // Index within the current chunk
+    var indexInCurrentChunkInMerged = 0;
+
+    diff.forEach(function(part) {
+
+      if (part.added) {
```

Comment:
Try rewriting with a 2-list strategy, shifting elements off the front of the diffs list onto the end of the merged list.

Reply:
Done! Looks great.